src/coordinatorapi.cpp

```diff
@@ -140,11 +146,19 @@ getCoordHostAndPort(CoordinatorMode mode, string *host, int *port)
     if (*host == "") {
       if (getenv(ENV_VAR_NAME_HOST)) {
         *host = getenv(ENV_VAR_NAME_HOST);
+        _cachedHost = getenv(ENV_VAR_NAME_HOST);
       } else if (getenv("DMTCP_HOST")) { // deprecated
         *host = getenv("DMTCP_HOST");
+        _cachedHost = getenv("DMTCP_HOST");
       } else {
         *host = DEFAULT_HOST;
+        _cachedHost = DEFAULT_HOST;
       }
+    } else {
+      // The caller's string object needs to be valid across
+      // multiple calls to this function, or else, the _cachedHost
+      // pointer will become a dangling pointer.
+      _cachedHost = host->c_str();
```

Comment:
And here.

Reply:
@karya0, It seems to me that what you're requesting should be the subject of a second PR or github issue.  This PR tries to fix a bug in which the hostname can be lost.  But it creates additional issues of concern (see below).  So, maybe it's better to separate out your concern as a github issue or a second PR to be addressed later by someone with more time available.

For the second proposed PR, we don't know `strlen(host->c_str())`.  So, in order to switch to `char[]` (actually to `char *`), we would need to do:
```
char *_cachedHost;
_cachedHost = JALLOC_HELPER_MALLOC(strlen(host->c_str())+1);
strcpy(_cachedHost, host->c_str());
```
This has its own dangers.  For example, there can be a small memory leak at this point in the code.  Of greater concern, `_cachedHost` is now pointing directly into the environment, and environment variables could possibly change during some later, orthogonal design change.  Of course, we could use `JALLOC_HELPER_MALLOC()` to copy environment strings, but then we create that small memory leak even in the standard case.

So when we're all done, it seems to me that we're just exchanging one hack (creating string objects in a constructor) for another hack (creating potential memory leaks using `char*` and `JALLOC_HELPER_MALLOC()`).

Reply:
Hi @gc00, There was some issue that probably @rohgarg and @jiajuncao looked at and decided to remove the static non-pod members. I don't remember the exact details but it had something to do with segfault during teardown.

Regarding the length of host string, we can do:
```char _cachedHost[HOST_NAME_MAX] = { 0 };```

Reply:
Didn't realize you inherited the static non-pod member from earlier. Approving this PR now. Will create another one to take care of the static aspect.